parachains/runtimes/collectives/collectives-polkadot/src/impls.rs

```diff
@@ -0,0 +1,51 @@
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::constants::POLKADOT_TREASURY_PALLET_INDEX;
+use frame_support::{
+	log,
+	traits::{Currency, Imbalance, OnUnbalanced, OriginTrait},
+};
+use sp_std::{boxed::Box, marker::PhantomData};
+use xcm::latest::{Fungibility, Junction, Parent};
+
+type NegativeImbalanceOf<T, I> = <<T as pallet_alliance::Config<I>>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::NegativeImbalance;
+
+type BalanceOf<T, I> = <<T as pallet_alliance::Config<I>>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::Balance;
+
+pub struct ToParentTreasury<T, I = ()>(PhantomData<(T, I)>);
+impl<T, I: 'static> OnUnbalanced<NegativeImbalanceOf<T, I>> for ToParentTreasury<T, I>
+where
+	T: pallet_xcm::Config + frame_system::Config + pallet_alliance::Config<I>,
+	BalanceOf<T, I>: Into<Fungibility>,
+{
+	fn on_unbalanced(amount: NegativeImbalanceOf<T, I>) {
+		let result = pallet_xcm::Pallet::<T>::teleport_assets(
+			<T as frame_system::Config>::Origin::root(),
```

Comment:
`teleport_assets` will try to withdraw them from some account. `Root` will pass any kind of `ensure_origin` checks, but I don't think it will have the `amount` in an account, since it's still an imbalance. So I _think_ it will fail on `WithdrawAsset`. @KiChjang ?

We could put something like this in:
```rust
fn on_unbalanced(amount: NegativeImbalanceOf<T, I>) {
    <pallet_balances::Pallet<R>>::resolve_creating(some_new_temp_account, &amount);
    let result = pallet_xcm::Pallet::<T>::teleport_assets(
        some_new_temp_account,
        ...
    );
    ...
}
```

Reply:
We first have to determine exactly which account it's withdrawing from, so let's do some investigation and look at the code in `pallet_xcm::do_teleport_assets` (because `teleport_assets` ultimately calls it):

https://github.com/paritytech/polkadot/blob/3b2b067d949c02ece20c76f1590066f2cd87b2b8/xcm/pallet-xcm/src/lib.rs#L825-L881

And there we see that there's a conversion from an `Origin` to a `MultiLocation` type in `T::ExecuteOrigin::ensure_origin()`, so now we'll need to see just how the collectives parachain defines `ExecuteXcmOrigin`:

https://github.com/paritytech/cumulus/blob/26301ad78a6dc2ac036517cfbd686bc27c139dc3/parachains/runtimes/collectives/collectives-polkadot/src/xcm_config.rs#L172

So now we look at the code for `EnsureXcmOrigin`:

https://github.com/paritytech/polkadot/blob/3b2b067d949c02ece20c76f1590066f2cd87b2b8/xcm/xcm-builder/src/origin_conversion.rs#L233-L259

Normally we'd have to dig through more to know what the `Conversion` type is, but since we know in advance that the `Origin` is root, we see from the code that it gets converted to `Here`. Neat.

Now, we know what the root origin `MultiLocation` is, but we don't know which account it gets converted to yet, so we then turn our attention to the code for executing `WithdrawAsset` for clues:

https://github.com/paritytech/polkadot/blob/3b2b067d949c02ece20c76f1590066f2cd87b2b8/xcm/xcm-executor/src/lib.rs#L274-L282

Now we can clearly see that the actual asset withdrawal code is done in `Config::AssetTransactor::withdraw_asset`, so now we look at what the collectives parachain has configured for its `AssetTransactor` and look at its `withdraw_asset` implementation:

https://github.com/paritytech/cumulus/blob/26301ad78a6dc2ac036517cfbd686bc27c139dc3/parachains/runtimes/collectives/collectives-polkadot/src/xcm_config.rs#L136

Aha, so it's a `CurrencyTransactor`, and it's defined here:

https://github.com/paritytech/cumulus/blob/26301ad78a6dc2ac036517cfbd686bc27c139dc3/parachains/runtimes/collectives/collectives-polkadot/src/xcm_config.rs#L63-L75

And finally we see the code for `withdraw_asset` defined in `CurrencyAdapter`:

https://github.com/paritytech/polkadot/blob/3b2b067d949c02ece20c76f1590066f2cd87b2b8/xcm/xcm-builder/src/currency_adapter.rs#L155-L164

The line that we're most interested in now is the following:

```rust
		let who =
			AccountIdConverter::convert_ref(who).map_err(|()| Error::AccountIdConversionFailed)?;
```

... because it converts a `MultiLocation` into an account that we can actually withdraw balances from, in which case, the next question is  "what is `AccountIdConverter`?" Looking at the definition for `CurrencyAdapter`, we see that it's the 3rd type parameter, and in the collectives parachain runtime, it's defined as `LocationToAccountId`, which is the following:

https://github.com/paritytech/cumulus/blob/26301ad78a6dc2ac036517cfbd686bc27c139dc3/parachains/runtimes/collectives/collectives-polkadot/src/xcm_config.rs#L51-L61

If we go to [`xcm-builder/src/location_conversion.rs`](https://github.com/paritytech/polkadot/blob/master/xcm/xcm-builder/src/location_conversion.rs), we'll see that none of the listed types here can help in converting a `Here` to a proper account ID, therefore, we can safely conclude that yes, `WithdrawAsset` will fail if the origin is root.

TL;DR: Joe is right in claiming that this code will fail during the execution of `WithdrawAsset`.

Reply:
fixed